README.md
# primary_secondary_progress_bar
Primary secondary progress bar, display 2 progress info in 1 progress bar

![Sample result](https://github.com/suesitran/primary_secondary_progress_bar/blob/master/image/primary_secondary_example.gif)

## Installation
Add primary_secondary_progress_bar as [dependencies into your pubspec.yaml](https://flutter.dev/docs/development/packages-and-plugins/using-packages)

## Example
```
Container(
                height: 300,
                padding: const EdgeInsets.all(8.0),
                child: PrimarySecondaryProgressBar(
                  context,
                  primaryLabel: "${_primaryMax.round()}%",
                  primaryMax: _primaryMax,
                  primaryValue: _primaryValue,
                  secondaryValue: _secondaryValue,
                  secondaryMax: _secondaryMax,
                  secondaryLabel: "${_secondaryValue.round()} days",
                  primaryIndicatorLine1: "${(_primaryMax - _primaryValue).round()}%",
                  primaryIndicatorLine2: "more",
                ),
              )
```


CHANGELOG.md
## 0.3.0

* Opt in to Null-NullSafety

## 0.2.0

* Support NullSafety

## 0.1.1

* Add Semantic label to support Talkback/voice over

## 0.1.0

* Release version 0.1
* Change Label widths to accommodate all label length

## 0.0.3

* Update Example Readme.md

## 0.0.2

* Update sample image

## 0.0.1

* Initial release of Primary Secondary progress bar.


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

//import 'package:flutter/material.dart';
//import 'package:flutter_test/flutter_test.dart';
//
//import 'package:primary_secondary_progress_bar_example/main.dart';
//
//void main() {
//  testWidgets('Verify Platform version', (WidgetTester tester) async {
//    // Build our app and trigger a frame.
//    await tester.pumpWidget(MyApp());
//
//    // Verify that platform version is retrieved.
//    expect(
//      find.byWidgetPredicate(
//        (Widget widget) => widget is Text &&
//                           widget.data.startsWith('Running on:'),
//      ),
//      findsOneWidget,
//    );
//  });
//}


test/primary_secondary_progress_bar_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const MethodChannel channel = MethodChannel('primary_secondary_progress_bar');

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });
}


lib/primary_secondary_progress_bar.dart
import 'package:flutter/material.dart';
import 'dart:ui' as ui;
import 'dart:math' as Math;

class PrimarySecondaryProgressBar extends StatelessWidget {
  final _PrimarySecondaryProgressBarPainter _painter;

  final sliderHeight;
  final sliderWidth;

  final semanticsLabel;

  PrimarySecondaryProgressBar(
    BuildContext context, {
    double primaryValue = 0.0,
    double primaryMax = 0.0,
    String primaryLabel = "",
    String primaryIndicatorLine1 = "",
    String primaryIndicatorLine2 = "",
    Color primaryTextColor = Colors.indigo,
    double secondaryValue = 0.0,
    double secondaryMax = 0.0,
    String secondaryLabel = "",
    Color secondaryTextColor = Colors.white,
    this.sliderWidth = 0.0,
    this.sliderHeight = 0.0,
    Color activeColor = Colors.white,
    double indicatorArrowHeight = 10.0,
    double indicatorHeight = 40.0,
    double indicatorWidth = 80.0,
    String semanticsLabel = "",
  }) : _painter = _PrimarySecondaryProgressBarPainter(context,
            primaryValue: primaryValue,
            primaryMax: primaryMax,
            primaryLabel: primaryLabel,
            primaryIndicatorLine1: primaryIndicatorLine1,
            primaryIndicatorLine2: primaryIndicatorLine2,
            primaryTextColor: primaryTextColor,
            secondaryMax: secondaryMax,
            secondaryValue: secondaryValue,
            secondaryLabel: secondaryLabel,
            secondaryTextColor: secondaryTextColor,
            activeColor: activeColor,
            indicatorArrowHeight: indicatorArrowHeight,
            indicatorHeight: indicatorHeight,
            indicatorWidth: indicatorWidth),
        semanticsLabel = semanticsLabel;

  @override
  Widget build(BuildContext context) {
    return Semantics(
      child: LayoutBuilder(
          builder: (context, constraint) => CustomPaint(
                size: constraint.biggest,
                painter: _painter
                  ..setSliderHeight(sliderHeight == 0.0 ? 50 : sliderHeight)
                  ..setSliderWidth(
                      sliderWidth == 0.0 ? constraint.maxWidth : sliderWidth),
              ),),
      label: semanticsLabel,
    );
  }
}

class _PrimarySecondaryProgressBarPainter extends CustomPainter {
  final BuildContext context;

  // primary bar
  final double primaryValue;
  final double primaryMax;
  final String primaryLabel;
  final String primaryIndicatorLine1;
  final String primaryIndicatorLine2;

  final double secondaryValue;
  final double secondaryMax;
  final String secondaryLabel;

  Color inactiveColor;
  final Color activeColor;
  final Color primaryTextColor;
  final Color secondaryTextColor;
  final double indicatorWidth;
  final double indicatorHeight;
  final double indicatorArrowHeight;

  late var _sliderHeight;
  late var _sliderWidth;

  _PrimarySecondaryProgressBarPainter(this.context,
      {this.primaryValue = 0,
      this.primaryMax = 0,
      this.primaryLabel = "",
      this.primaryIndicatorLine1 = "",
      this.primaryIndicatorLine2 = "",
      this.primaryTextColor = Colors.white,
      this.secondaryValue = 0,
      this.secondaryMax = 100,
      this.secondaryLabel = "",
      this.secondaryTextColor = Colors.blue,
      this.activeColor = Colors.blue,
      this.indicatorWidth = 40,
      this.indicatorHeight = 50,
      this.indicatorArrowHeight = 50}) : this.inactiveColor = activeColor.withOpacity(0.7);

  void setSliderHeight(double height) => this._sliderHeight = height;
  void setSliderWidth(double width) => this._sliderWidth = width;

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }

  @override
  void paint(ui.Canvas canvas, ui.Size size) {
    var top = (size.height - _sliderHeight) / 2;
    var bottom = top + _sliderHeight;
    var fontSize = Theme.of(context).textTheme.bodyText1?.fontSize ?? 18;

    var sliderRect = ui.Rect.fromLTRB(0, top, _sliderWidth, bottom);

    _drawSecondaryIndicator(canvas, size, sliderRect);
    _drawSlider(canvas, size, sliderRect, fontSize);
  }

  void _drawSlider(
      ui.Canvas canvas, ui.Size size, ui.Rect sliderRect, double fontSize) {
    var progressRatio = primaryMax == 0 ? 0 : primaryValue / primaryMax;

    // draw slider
    ui.Paint sliderBackground = new ui.Paint()
      ..color = inactiveColor
      ..strokeCap = ui.StrokeCap.round
      ..style = ui.PaintingStyle.fill
      ..isAntiAlias = true;
    ui.Paint sliderProgress = new ui.Paint()
      ..color = activeColor
      ..strokeCap = ui.StrokeCap.round
      ..style = ui.PaintingStyle.fill
      ..isAntiAlias = true;
    final progressRadius = ui.Radius.circular(10.0);

    canvas.drawRRect(ui.RRect.fromRectAndRadius(sliderRect, progressRadius),
        sliderBackground);

    var progressWidth = sliderRect.width * progressRatio;
    canvas.drawRRect(
        ui.RRect.fromRectAndRadius(
            ui.Rect.fromLTRB(sliderRect.left, sliderRect.top, progressWidth,
                sliderRect.bottom),
            progressRadius),
        sliderProgress);

    // draw primary label => max primary
    TextPainter sliderPainter = new TextPainter(
        text: TextSpan(
            text: primaryLabel, style: TextStyle(color: primaryTextColor)),
        textDirection: TextDirection.ltr,
        textAlign: TextAlign.right,
        maxLines: 1);
    final sliderTextRect = Rect.fromLTRB(sliderRect.left + 5, sliderRect.top,
        sliderRect.right - 5, sliderRect.bottom);
    sliderPainter.layout(maxWidth: sliderTextRect.width);
    drawText(canvas, sliderTextRect, sliderPainter, textAlign: TextAlign.right);

    // draw expenses indicator
    final arrowWidth = indicatorWidth / 4;

    // define arrow point offsets
    var arrowCenter = progressWidth;
    var arrowLeft = arrowCenter - arrowWidth / 2;
    var arrowRight = arrowCenter + arrowWidth / 2;

    // calculate actual indicatorWidth base on text width, and use this indicator width to draw
    final line1Painter = new TextPainter(
        text: TextSpan(
            text: primaryIndicatorLine1,
            style: TextStyle(fontSize: fontSize, color: primaryTextColor)),
        textDirection: TextDirection.ltr,
        textAlign: TextAlign.center,
        maxLines: 1)
      ..layout(minWidth: indicatorWidth);
    final line2Painter = new TextPainter(
        text: TextSpan(
            text: primaryIndicatorLine2,
            style:
                TextStyle(fontSize: fontSize * 0.6, color: primaryTextColor)),
        textDirection: TextDirection.ltr,
        textAlign: TextAlign.center,
        maxLines: 1)
      ..layout(minWidth: indicatorWidth);

    var actualIndicatorWidth =
        Math.max(line1Painter.width, line2Painter.width) +
            5 /* padding left */ +
            5 /* padding right*/;

    var indicatorLeft = progressWidth - actualIndicatorWidth / 2;
    if (indicatorLeft <= 0) {
      // adjust bubble when bubble is overlapping left corner
      if (progressWidth < arrowWidth / 2) {
        indicatorLeft = progressWidth;
      } else {
        indicatorLeft = 0;
      }
      arrowLeft = progressWidth;
    }
    var indicatorRight = indicatorLeft + actualIndicatorWidth;
    if (indicatorRight >= size.width) {
      // adjust bubble when bubble is overlapping right conner
      if (progressWidth + actualIndicatorWidth >= size.width) {
        indicatorRight = progressWidth;
        indicatorLeft = size.width - actualIndicatorWidth;
        arrowRight = arrowCenter;
      } else {
        indicatorRight = size.width - arrowWidth / 2;
        arrowRight = arrowCenter;
      }
      indicatorLeft = indicatorRight - actualIndicatorWidth;
    }
    var indicatorBottom = sliderRect.top - indicatorArrowHeight;
    var indicatorTop = indicatorBottom - indicatorHeight;

    final indicatorRect = ui.Rect.fromLTRB(
        indicatorLeft, indicatorTop, indicatorRight, indicatorBottom);
    canvas.drawRRect(ui.RRect.fromRectAndRadius(indicatorRect, progressRadius),
        sliderProgress);

    // create little arrow path
    ui.Path path = new ui.Path()
      ..moveTo(arrowLeft, indicatorBottom - indicatorHeight / 2)
      ..lineTo(arrowLeft, indicatorBottom)
      ..lineTo(arrowCenter, indicatorBottom + indicatorArrowHeight)
      ..lineTo(arrowRight, indicatorBottom)
      ..lineTo(arrowRight, indicatorBottom - indicatorHeight / 2);
    canvas.drawPath(path, sliderProgress);

    // draw primary indicator string
    if (primaryIndicatorLine2.isEmpty) {
      drawText(canvas, indicatorRect, line1Painter);
    } else {
      draw2LinesText(canvas, indicatorRect, line1Painter, line2Painter);
    }
  }

  void _drawSecondaryIndicator(
      ui.Canvas canvas, ui.Size size, ui.Rect sliderRect) {
    final secondaryTextMarginTop = 10.0;

    ui.Paint dayIndicator = new ui.Paint()
      ..color = activeColor
      ..strokeCap = ui.StrokeCap.round
      ..style = ui.PaintingStyle.stroke
      ..strokeWidth = 3.0
      ..isAntiAlias = true;

    var daysRatio = secondaryMax > 0 ? secondaryValue / secondaryMax : 0;
    num indicatorPosition = daysRatio * sliderRect.width;
    canvas.drawLine(
        ui.Offset(indicatorPosition as double, sliderRect.top - secondaryTextMarginTop),
        ui.Offset(
            indicatorPosition, sliderRect.bottom + secondaryTextMarginTop),
        dayIndicator);

    var textAlign = ui.TextAlign.center;

    final textPainter = new TextPainter(
        text: TextSpan(
            text: secondaryLabel, style: TextStyle(color: secondaryTextColor)),
        textDirection: TextDirection.ltr,
        textAlign: textAlign,
        maxLines: 1);
    textPainter.layout();

    final secondaryTextWidth = textPainter.width;

    num textLeft = indicatorPosition - secondaryTextWidth / 2;
    if (textLeft <= 0) {
      textLeft = 0.0;
      textAlign = ui.TextAlign.start;
    }
    num textRight = textLeft + secondaryTextWidth;

    if (textRight >= size.width) {
      textRight = size.width;
      textAlign = ui.TextAlign.end;
      textLeft = textRight - secondaryTextWidth;
    }
    final textTop = sliderRect.bottom + secondaryTextMarginTop;
    final textBottom = textTop + textPainter.height;

    final textRect = ui.Rect.fromLTRB(textLeft as double, textTop, textRight as double, textBottom);

    drawText(canvas, textRect, textPainter);
  }

  /// Draw 2 lines text, to be used to draw primary text layout
  /// Text Alignment is always Center
  void draw2LinesText(ui.Canvas canvas, ui.Rect boundary,
      TextPainter line1Painter, TextPainter line2Painter) {
    ui.Offset primaryLine = new ui.Offset(0.0, boundary.height * 0.05);
    ui.Offset secondaryLine = new ui.Offset(0.0, boundary.height * 0.6);

    // draw line 1
    canvas.save();
    canvas.translate(boundary.left + (boundary.width - line1Painter.width) / 2,
        boundary.top);
    line1Painter.paint(canvas, primaryLine);
    canvas.restore();

    // draw line1
    canvas.save();
    canvas.translate(boundary.left + (boundary.width - line2Painter.width) / 2,
        boundary.top);
    line2Painter.paint(canvas, secondaryLine);
    canvas.restore();
  }

  /// Draw single line text
  /// Text Alignment by default is center
  void drawText(ui.Canvas canvas, ui.Rect boundary, TextPainter painter,
      {TextAlign textAlign = TextAlign.center}) {
    ui.Offset offset = new ui.Offset(
        0.0, boundary.top + (boundary.height - painter.height) / 2);

    canvas.save();
    switch (textAlign) {
      case TextAlign.center:
        canvas.translate(
            boundary.left + (boundary.width - painter.width) / 2, 0.0);
        break;
      case TextAlign.start:
      case TextAlign.left:
      case TextAlign.justify:
        break;
      case TextAlign.end:
      case TextAlign.right:
        canvas.translate(boundary.left + (boundary.width - painter.width), 0.0);
        break;
    }
    painter.paint(canvas, offset);
    canvas.restore();
  }
}


ios/Classes/PrimarySecondaryProgressBarPlugin.m
#import "PrimarySecondaryProgressBarPlugin.h"

@implementation PrimarySecondaryProgressBarPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"primary_secondary_progress_bar"
            binaryMessenger:[registrar messenger]];
  PrimarySecondaryProgressBarPlugin* instance = [[PrimarySecondaryProgressBarPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  } else {
    result(FlutterMethodNotImplemented);
  }
}

@end


android/src/main/java/com/example/primary_secondary_progress_bar/PrimarySecondaryProgressBarPlugin.java
package com.example.primary_secondary_progress_bar;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** PrimarySecondaryProgressBarPlugin */
public class PrimarySecondaryProgressBarPlugin implements MethodCallHandler {
  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "primary_secondary_progress_bar");
    channel.setMethodCallHandler(new PrimarySecondaryProgressBarPlugin());
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:primary_secondary_progress_bar/primary_secondary_progress_bar.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: _PrimarySecondaryExample(),
    );
  }
}

class _PrimarySecondaryExample extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return _PrimarySecondaryExampleState();
  }
}

class _PrimarySecondaryExampleState extends State<_PrimarySecondaryExample> {

  final _primaryMax = 100.0;
  double _primaryValue = 0.0;

  final _secondaryMax = 30.0;
  double _secondaryValue = 0.0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: const Text('Progress Bar Example'),
        ),
        body: Column(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Card(
              color: Colors.indigo,
              borderOnForeground: true,
              child: Container(
                height: 300,
                padding: const EdgeInsets.all(8.0),
                child: PrimarySecondaryProgressBar(
                  context,
                  primaryLabel: "${_primaryMax.round()}%",
                  primaryMax: _primaryMax,
                  primaryValue: _primaryValue,
                  secondaryValue: _secondaryValue,
                  secondaryMax: _secondaryMax,
                  secondaryLabel: "${_secondaryValue.round()} days",
                  primaryIndicatorLine1: "${(_primaryMax - _primaryValue).round()}%",
                  primaryIndicatorLine2: "more",
                ),
              ),
            ),
            FittedBox(
              child: Column(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                crossAxisAlignment: CrossAxisAlignment.start,
                children: <Widget>[
                  Text("Primary Value", style: Theme.of(context).textTheme.bodyText2,),
                  Slider(
                    value: _primaryValue,
                    min: 0,
                    max: _primaryMax,
                    onChanged: (value) => setState(() => _primaryValue = value),
                  ),
                  Text("Secondary value", style: Theme.of(context).textTheme.bodyText2,),
                  Slider(
                    value: _secondaryValue,
                    min: 0.0,
                    max: _secondaryMax,
                    onChanged: (value) => setState(() => _secondaryValue = value),
                  )
                ],
              ),
            )
          ],
        ));
  }
}
